ostd/src/mm/page/mod.rs

```diff
@@ -52,24 +53,36 @@ impl<M: PageMeta> Page<M> {
     ///
     /// The function panics if:
     ///  - the physical address is out of bound or not aligned;
-    ///  - the page is already in use.
+    ///  - the page is already in use;
+    ///  - the metadata's size exceeds [`PAGE_METADATA_MAX_SIZE`];
+    ///  - the alignment of the metadata is more than 8 bytes.
     pub fn from_unused(paddr: Paddr, metadata: M) -> Self {
         assert!(paddr % PAGE_SIZE == 0);
         assert!(paddr < MAX_PADDR.load(Ordering::Relaxed) as Paddr);
+
+        // Checking unsafe preconditions of the `PageMeta` trait.
+        debug_assert!(size_of::<M>() <= PAGE_METADATA_MAX_SIZE);
+        debug_assert!(align_of::<M>() <= PAGE_METADATA_MAX_ALIGN);
```

Comment:
Is it wrong/unsound to do this as `assert!` (in `const`) and make the `PageMeta` trait safe?

```rust
let _ = const {
    assert!(size_of::<M>() <= PAGE_METADATA_MAX_SIZE);
    assert!(align_of::<M>() <= PAGE_METADATA_MAX_ALIGN);
};
```
(I'm not sure if `const_assert!` can be used here, but I'm sure `assert!` in the `const` block will compile).

Reply:
The proposed way won't compile:

```
error: overly complex generic constant
  --> /root/asterinas/ostd/src/mm/page/mod.rs:64:15
   |
64 |         const { assert!(size_of::<M>() <= PAGE_METADATA_MAX_SIZE )};
   |               ^^--------------------------------------------------^
   |                 |
   |                 control flow is not supported in generic constants
   |
   = help: consider moving this anonymous constant into a `const` function

error: overly complex generic constant
  --> /root/asterinas/ostd/src/mm/page/mod.rs:64:23
   |
64 |           let _ = const {
   |  _______________________^
65 | |     assert!(size_of::<M>() <= PAGE_METADATA_MAX_SIZE);
66 | |     assert!(align_of::<M>() <= PAGE_METADATA_MAX_ALIGN);
67 | | };
   | |_^ blocks are not supported in generic constants
   |
   = help: consider moving this anonymous constant into a `const` function
   = note: this operation may be supported in the future
```

There exist a version that compiles:

```rust
assert!(const { size_of::<M>() <= PAGE_METADATA_MAX_SIZE });
```

It is neither wrong nor unsound to make const assertions in the latter way. However, if we just rely on the assertion here, incompatible `PageMeta` implementations will trigger only runtime errors rather than compile errors. I believe that since the layout of the structs can be statically decided, the error should be compile errors.

Reply:
OK, I see. "overly complex generic constant" is an annoying error.

> I believe that since the layout of the structs can be statically decided, the error should be compile errors.

Yes, I agree. Then the current implementation is reasonable to me.